draft-ietf-tls-tls13.md

```diff
@@ -2089,28 +2088,42 @@ selected_group
-generate a correct ClientHello pair.
-
-Upon receipt of a HelloRetryRequest, the client MUST first verify
-that the "selected_group" field does not identify a group which
-was not in the original ClientHello. If it was present, then
-the client MUST abort the handshake with a fatal "handshake_failure"
-alert. Clients SHOULD also abort with "handshake_failure" in response to any second
-HelloRetryRequest which was sent in the same connection (i.e.,
-where the ClientHello was itself in response to a HelloRetryRequest).
-
-Otherwise, the client MUST send a ClientHello with a new
-ClientKeyShare extension to the server. The ClientKeyShare MUST append
-a new ClientKeyShareOffer which is consistent with the
-"selected_group" field to the groups in the original ClientKeyShare.
-
-Upon re-sending the ClientHello and receiving the
-server's ServerHello/ServerKeyShare, the client MUST verify that
-the selected CipherSuite and NamedGroup match that supplied in
-the HelloRetryRequest.
+The server_version, cipher_suite, and extensions fields have the same
+meanings as their corresponding values in the ServerHello. The server
+SHOULD send only the extensions necessary for the client to generate a
+correct ClientHello pair. As with a ServerHello, only extensions first
+offered by the client can appear in a HelloRetryRequest.
+
+Upon receipt of a HelloRetryRequest, the client MUST first verify that
+the selected_group was offered in its "supported_groups" extension and
+that no ClientKeyShareOffer for the selected_group was offered in its
+"client_key_share" extension. If either of these conditions are detected,
+then the client MUST abort the handshake with a fatal "handshake_failure"
+alert.
+
+Clients SHOULD also abort with an "handshake_failure" alert in response
+to any second HelloRetryRequest which is sent on the same connection
+(i.e., where the ClientHello was itself in response to a
+HelloRetryRequest).
+
+If the HelloRetryRequest is verified as acceptable, the client MUST send
+a new ClientHello with an updated ClientKeyShare extension to the server.
+This ClientKeyShare MUST contain the exact previous set of offered keys
+with a new ClientKeyShareOffer value for the server's selected_group
+appended to the end of the list (or a new extension with exactly one
+value, if none was provided initially). All other fields in this new
+ClientHello MUST match the first attempt. Note that omitting any
+information previously included in the initial ClientHello introduces
+the risk of downgrade attack, as this retry exchange is unauthenticated.
+
+After sending the new ClientHello and receiving a new ServerHello,
+the client MUST verify that the same ProtocolVersion, CipherSuite, and
+NamedGroup specified in the server's HelloRetryRequest are selected in the
+server's new ServerHello/ServerKeyShare. If any of these values differ,
+the client MUST abort the connection with a fatal "handshake_failure"
+alert.
```

Comment:
On Monday, October 05, 2015 02:37:42 pm ekr wrote:
> > +After sending the new ClientHello and receiving a new ServerHello,
> > +the client MUST verify that the same ProtocolVersion, CipherSuite, and
> > +NamedGroup specified in the server's HelloRetryRequest are selected in the
> > +server's new ServerHello/ServerKeyShare. If any of these values differ,
> > +the client MUST abort the connection with a fatal "handshake_failure"
> > +alert.
> 
> This seems to be inconsistent about how to name these fields. "server_version" above and ProtocolVersion here. Yes, this is preexisting, but we should make them match

Sure. I amended the last commit for changes for your comments to change this too & repushed.